errorVars/errorVars.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package errorVars
+
+import (
+	"go/ast"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+const (
+	appErrorString = "*github.com/mattermost/mattermost-server/v5/model.AppError"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "errorVars",
+	Doc:  "check for non valid type assignations to err and appErr prefixed variables",
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(node ast.Node) bool {
+			switch x := node.(type) {
+			case *ast.AssignStmt:
+				for idx, lhsExpr := range x.Lhs {
```

Comment:
I am a bit confused. I thought we were only checking for variable names? I am not sure why assigning `*model.AppError` to `error` is wrong. Any struct satisfying the `error` interface can be an error value.

I thought we were going to check that variable _names_ should be err for error types, and `appErr` for `*model.AppError` types.

Reply:
There are 2 different checks.

We can discard this one, but this one is about not generating the problem of having pointers to *model.AppError in error variables that the assertions/ifs works in the wrong way.

Reply:
Ah gotcha, so https://github.com/mattermost/mattermost-govet/pull/18 is the one which I was thinking.

So I think what you are referring to is https://golang.org/doc/faq#nil_error where a variable of type `*model.AppError` but of nil value, will always by non-nil. That will happen due to incorrectly written code and we should fix the code. But assigning a `*model.AppError` struct to error is fine.

Reply:
Yes, it is not necessarily wrong, but in general, we don't do that in our code, and in the past has bitten us. This is not any kind of `*model.AppError` coercion to `error`, it is just for previous existing `error` type variables reused to store a `*model.AppError` variable. That usually we do accidentally, not on purpose.

The other case that is more common and is not check by this is "returning `*model.AppError` as `error`" what is perfectly fine, and should be the way to do it.

Again, this is going to be probably fixed by the "naming convention" anyway, so... we can focus our effors on the other check and close this one.

Reply:
Ohh I see now. Yes sure, I am fine with that. I thought this was checking the  "returning *model.AppError as error"  case which is completely valid.

It would have been helpful if the PR message had some examples with a bit more explanation to avoid this confusion. :)

Reply:
:+1: